### PR TITLE
Avoid reporting empty debug info ranges to the vm

### PR DESCRIPTION
**Pull request by BrianBohe**
This PR is intended to solve issue #47202.

The thing is empty variable live ranges are useless. We are reporting them because they can get extended later if the variable becomes alive in the immediately next emitted instruction. If an empty live range is not getting extended, which we can realize after emitting all the code or creating a new live range for the same variable, we can remove it.

Tested running
```
<path_to_repo>\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\corerun.exe  <path_to_repo>\artifacts\bin\coreclr\windows.x64.Debug\R2RDump\R2RDump.dll --in  <path_to_repo>\artifacts\bin\coreclr\windows.x64.Debug\System.Private.CoreLib.dll > output.txt
```
before and after this PR, checking the differences manually. Files are big and I wasn't able to get a nice diff that I could post here. I am open to any suggestions for more testing.

### CODE REVIEW

**Comment by BrianBohe** [comment]
I am running now R2R dump with https://github.com/BrianBohe/runtime/commit/ebb0a4519d07b04caf813a07e930fa8a1b0c0fc2 and found there are still empty variable ranges. Will debug it a little more and open this PR again.

**Comment by BrianBohe** [comment]
I found a case where a variable dies at the beginning of a basic block, and becomes alive before an instruction is emitter. We would want to extend the live ranges in this case:

```
                                                                        /--*  t52    long   
Generating: N087 ( 10, 16) [000054] DA---------                         *  STORE_LCL_VAR long   V06 loc5         d:3 rcx REG rcx
							V06 in reg rcx is becoming live  [000054]
							Live regs: 00000000 {} => 00000002 {rcx}
							Live vars: {} => {V06}
New debug range: first
IN000f:        jmp      L_M3923_BB06

Variable Live Range History Dump for BB04
V04 loc3: rdx [(G_M3923_IG03,ins#5,ofs#27), (G_M3923_IG04,ins#1,ofs#10)]
V06 loc5: rcx [(G_M3923_IG04,ins#2,ofs#13), ...]

=============== Generating BB05 [060..06B), preds={BB01} succs={BB06} flags=0x00000000.20010020: i label LIR 
BB05 IN (1)={V09    }
     OUT(1)={    V06}

Recording Var Locations at start of BB05
  V09(rcx)
Change life 0000000000000008 {V06} -> 0000000000000002 {V09}
							V06 in reg rcx is becoming dead  [------]
							Live regs: (unchanged) 00000000 {}
							V09 in reg rcx is becoming live  [------]
							Live regs: 00000000 {} => 00000002 {rcx}
							Live regs: (unchanged) 00000002 {rcx}
							GC regs: (unchanged) 00000000 {}
							Byref regs: (unchanged) 00000000 {}

      L_M3923_BB05:

      G_M3923_IG04:        ; offs=000044H, funclet=00, bbWeight=0.50, byref
Mapped BB05 to G_M3923_IG05
Label: IG05, GCvars=0000000000000000 {}, gcrefRegs=00000000 {}, byrefRegs=00000000 {}

Scope info: begin block BB05, IL range [060..06B)
Added IP mapping: 0x0061 STACK_EMPTY (G_M3923_IG05,ins#0,ofs#0) label
Generating: N091 (???,???) [000093] -----------                            IL_OFFSET void   INLRT @ 0x061[E-] REG NA
Generating: N093 (  1,  1) [000086] -----------                   t86 =    LCL_VAR   long   V09 cse0         u:1 rcx (last use) REG rcx <l:$100, c:$140>
                                                                        /--*  t86    long   
Generating: N095 (  5,  4) [000016] DA--G------                         *  STORE_LCL_VAR long   V06 loc5         d:2 rcx REG rcx
							V09 in reg rcx is becoming dead  [000086]
							Live regs: 00000002 {rcx} => 00000000 {}
							Live vars: {V09} => {}
							V06 in reg rcx is becoming live  [000016]
							Live regs: 00000000 {} => 00000002 {rcx}
							Live vars: {} => {V06}
New debug range: new var or location
```

**Comment by BrianBohe** [comment]
SPMI results show that this approach and [its alternative](https://github.com/dotnet/runtime/pull/77370)

- has no significant change on asm diff and downgrade throughput on average +0.01~+0.03 %.
- This PR is consistently improving 2x in the good cases compared to [its alternative](https://github.com/dotnet/runtime/pull/77370). Improving values are between -0.00% and -0.02 %.
- The [alternative](https://github.com/dotnet/runtime/pull/77370) is consistently doing 2x worse in the bad cases compared to this PR. Downgrading values are mostly between +0.00% and +0.03 %.
- The worst bad case is consistently a MinOpt collection called "libraries.crossgen2.<OS>.<arch>.checked.mch", which throughput is +0.04~+0.07%  for this PR and +0.08~+0.16 % the alterantive.

I was not able to get prints of my code on the log using SPMI locally. But running crossgen to build System.private.corelib dll gets:

- 32322 functions jitted.
- 189939 variable live ranges created on jit memory. 189939/32322 = 5.8 live ranges per function.
- Of which 18119 are empty and not reported (after this changes). 18119/32322 = 0.56 empty live ranges per function.
- A reported range instance (the struct used to communicate with the vm) uses 40 bytes, so 724760 bytes are requested to the vm and lost. 724760/32322 = 22.42 lost bytes per jitted function.

Obs: ~10% of variable live ranges are empty after code is generated and could not be omitted before computing emitLocation absolute offsets.

On main branch, we are requesting memory to the vm, using it entirely, but reporting empty ranges. Running the same command I got:
- 208763 variable live ranges created on jit memory. 6.45 live ranges per function.
- Of which 32522 are empty and not reported (after this changes). 1.006 empty live ranges per function.
- 1300880  bytes are requested to the vm and lost. 40.24 lost bytes per jitted function.

Obs: ~15.57% of variable live ranges are empty. This PR introduces a change which allows extending more variable live ranges and replace the ones we know are empty before finishing emitting code, that is why we see a difference in the % of total variable live ranges.

**Comment by BrianBohe** [comment]
The important aspect of this pr and its alternative approach is to decide whether we want to filter empty live ranges before asking the vm for memory or not. I am moving the updated logic [extending the variable range](https://github.com/dotnet/runtime/pull/77439) and the [removal of duplicated prints](https://github.com/dotnet/runtime/pull/77447) to other PRs. After deciding between this PR and [its alternative](https://github.com/dotnet/runtime/pull/77370), I will open a new PR to decide whether we want to preemptively remove variable live ranges when we can tell from its instruction group and offset that it is empty, before finishing emitting code. 

@jakobbotsch pointed me that the memory won't be lost as the vm is taking it, comprising and releasing it.

**Comment by BrianBohe** [comment]
Update now that we are just comparing both ideas:

- No ASM diff on both PRs
- No significant throughput differences found on this PR. A few small regressions happen on the alternative +%0.00~+%0.09.

Giving the fact that the extra memory is being free after the jit, I see no point on choosing the alternative.

Links to [this spmi diff ](https://dev.azure.com/dnceng-public/public/_build/results?buildId=63056&view=ms.vss-build-web.run-extensions-tab) and [the alternative](https://dev.azure.com/dnceng-public/public/_build/results?buildId=63085&view=ms.vss-build-web.run-extensions-tab)

**Review by jakobbotsch — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Looks like the VM is not happy that we may end up reporting no mappings after asking for memory. I think it should be fine to fix the VM side to handle this correctly.

**Comment by jakobbotsch** [comment]
It looks like crossgen2 will need a small fix to properly free the array:

https://github.com/dotnet/runtime/blob/cac1b598b666b126069d9ca651b6429673e13f57/src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs#L1303-L1322

NAOT looks ok.

**Comment by jakobbotsch** [comment]
It might be better to make sure we don't call back into the VM with "valid pointer" + "0 length" combo so that we don't change the contract here. E.g. something like:

```diff
--- a/src/coreclr/jit/ee_il_dll.cpp
+++ b/src/coreclr/jit/ee_il_dll.cpp
@@ -694,7 +694,19 @@ void Compiler::eeSetLVdone()
     }
 #endif // DEBUG
 
-    info.compCompHnd->setVars(info.compMethodHnd, eeVarsCount, (ICorDebugInfo::NativeVarInfo*)eeVars);
+    if (eeVarsCount > 0)
+    {
+        info.compCompHnd->setVars(info.compMethodHnd, eeVarsCount, (ICorDebugInfo::NativeVarInfo*)eeVars);
+    }
+    else
+    {
+        info.compCompHnd->setVars(info.compMethodHnd, 0, nullptr);
+
+        if (eeVars != nullptr)
+        {
+            info.compCompHnd->freeArray(eeVars);
+        }
+    }
 
     eeVars = nullptr; // We give up ownership after setVars()
 }
```

**Comment by BrianBohe** [comment]
@jakobbotsch thanks for the review

File: src/coreclr/jit/ee_il_dll.cpp [hunk 1]
```diff
@@ -693,8 +693,14 @@ void Compiler::eeSetLVdone()
         eeDispVars(info.compMethodHnd, eeVarsCount, (ICorDebugInfo::NativeVarInfo*)eeVars);
     }
 #endif // DEBUG
-
-    info.compCompHnd->setVars(info.compMethodHnd, eeVarsCount, (ICorDebugInfo::NativeVarInfo*)eeVars);
+    if (0 < eeVarsCount)
+    {
+        info.compCompHnd->setVars(info.compMethodHnd, eeVarsCount, (ICorDebugInfo::NativeVarInfo*)eeVars);
```
Comment (jakobbotsch) [inline src/coreclr/jit/ee_il_dll.cpp:698]: This is changing the contract of this JIT-EE function, before this we were calling `setVars(info.compMethodHnd, 0, nullptr)` in the case when there are no variable live ranges, through this path:

https://github.com/dotnet/runtime/blob/689350960f9cbb76c7aa469f129b5e7227324dae/src/coreclr/jit/codegencommon.cpp#L6813-L6819

File: src/coreclr/jit/ee_il_dll.cpp [hunk 2]
```diff
@@ -693,9 +693,15 @@ void Compiler::eeSetLVdone()
         eeDispVars(info.compMethodHnd, eeVarsCount, (ICorDebugInfo::NativeVarInfo*)eeVars);
     }
 #endif // DEBUG
+    if (0 == eeVarsCount && eeVars != nullptr)
```
Comment (jakobbotsch) [inline src/coreclr/jit/ee_il_dll.cpp:696]: Nit: Our [coding conventions](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md) specify that subexpressions should be fully parenthesized and that constants should (usually) be on the RHS, like this:

```suggestion
    if ((eeVarsCount == 0) && (eeVars != nullptr))
```

**Review by jakobbotsch — APPROVED** [review]

**Comment by cshung** [comment]
> I am running now R2R dump with [BrianBohe@ebb0a45](https://github.com/BrianBohe/runtime/commit/ebb0a4519d07b04caf813a07e930fa8a1b0c0fc2) and found there are still empty variable ranges. Will debug it a little more and open this PR again.

Would you mind adding the validation that we do not report empty ranges to [here](https://github.com/dotnet/runtime/blob/92971c9ae60b40548ec1dda54de977afc3ff8021/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs#L1172)? This will make sure it stays correct even when the JIT changes.

**Comment by BrianBohe** [comment]
> Would you mind adding the validation that we do not report empty ranges to [here](https://github.com/dotnet/runtime/blob/92971c9ae60b40548ec1dda54de977afc3ff8021/src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs#L1172)? This will make sure it stays correct even when the JIT changes.

Sure, Who and how is running that function currently? Is there a job scheduled running this once in a while or for every commit in PRs?

**Pull request merged**